Remove HAVE_RAM_AVAILABLE_128K in favour of dynamic heap checking in test

stevew817 opened the pull request:
## Description
Ref. email discussions with @ronald-cron-arm and @gilles-peskine-arm. `HAVE_RAM_AVAILABLE_128K` was put in as a hasty fix, and doesn't scale that well (i.e. what about 64K? 32K?). It's definition was also pretty unclear: is it 128K total? Or 128K heap size?

Since the test system now got extended with `ASSERT_ALLOC_WEAK`, we can leverage that specifically for those tests which try to exercise the absolute largest key size possible. If allocating a buffer the size of the key fails, then the test is skipped, instead of requiring the user to know about a well-hidden compile-time define that needed to be commented out manually in the test files.

## Status
**READY**

## Requires Backporting
NO  

## Migrations
NO

## Todos
- [X] Tests
- [X] Documentation

## Steps to test or reproduce
No impact on testing. All tests should still succeed on regular targets (with large enough memory sizes). This change enables on-target testing for constrained devices with the same set of tests as otherwise run by default.

daverodgman reviewed (approved):
LGTM

gilles-peskine-arm commented on tests/suites/test_suite_psa_crypto.function:
```diff
@@ -5321,6 +5314,26 @@ exit:
+
+    /* Check that the target running this test can accomodate large
+     * keys on its heap, before calling the actual generate_key test */
+    ASSERT_ALLOC_WEAK(buffer, key_bytes);
```
This is fragile and I fear that we'll run into problems in practice if the heap just barely has room for the key, but not for the key plus other data that the test allocates. Then the buffer allocation would succeed here, but the test would run out of memory later.

I would prefer to go through the test, and in the large-key case, allow the function that's supposed to allocate a large amount of memory to return `PSA_ERROR_OUT_OF_MEMORY.

gilles-peskine-arm replied:
Actually there's a case where this definitely won't work: on a target where crypto runs in a separate service, this checks whether the client has room on its heap, but what matters for the test to pass is that the service has enough room. So please change to actually running the test, and in the “large” case, allowing `PSA_ERROR_OUT_OF_MEMORY`.

This is also better because it gives us a chance to catch errors in out-of-memory cleanup.

stevew817 replied:
Done.

stevew817 commented:
@ronald-cron-arm Since I addressed all comments, can I remove the 'needs work' label?

ronald-cron-arm commented:
> 
> 
> @ronald-cron-arm Since I addressed all comments, can I remove the 'needs work' label?

Yes please.

daverodgman reviewed (approved):
Changes address Gilles' comments - LGTM

gilles-peskine-arm commented:
There's a merge conflict in `test_suite_psa_crypto.data`, and there'll be a conflict with https://github.com/ARMmbed/mbedtls/pull/4012 as well. Since #4012 is more complex, may I go first?

stevew817 commented:
@gilles-peskine-arm If you think #4012 will be merged soon, then sure. I'd like to close this one out rather quickly, too.

stevew817 commented:
@daverodgman and @gilles-peskine-arm I've rebased and updated the PR. Please review.

Also, I mistakenly pushed to origin instead of my fork. Sorry about that. Deleted the branch immediately after, but it still triggered the Travis 'Branch' job, which will obviously fail now that the branch is deleted. So it can be ignored safely.

gilles-peskine-arm commented:
> Also, I mistakenly pushed to origin instead of my fork.

This is what I use in my `.git/config` to avoid accidental pushes:
```
[remote "upstream-public"]
        url = https://github.com/ARMmbed/mbedtls.git
        fetch = +refs/heads/*:refs/remotes/upstream-public/*
        fetch = +refs/pull/*/head:refs/remotes/upstream-public/pr/*
        pushurl = do not push to this remote
```

stevew817 commented:
Rebased due to merge conflict with #3933. Please re-approve @gilles-peskine-arm and @daverodgman

gilles-peskine-arm reviewed (approved):
This is a rebase of the version I previously approved (951ce6cb16997aa83a5bbbb64cbba90580cbc192) on top of the tip of the target branch (00c3e87422ebb0de9b60584d00e129f0d5024687). I did a merge and ended up with identical contents.

ronald-cron-arm commented:
@daverodgman is the latest version still ok with you ?

daverodgman reviewed (approved):
Looks good, thanks Steven
